src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs

```diff
@@ -70,8 +70,41 @@ public override void OnStartConnect()
 
                 PeerAddress peerAddress = this.peerAddressManager.FindPeer(ipEndpoint);
                 if (peerAddress != null && !this.IsPeerConnected(peerAddress.Endpoint))
-                    await ConnectAsync(peerAddress).ConfigureAwait(false);
+                {
+                    // Nodes disallow connection to peers in same range to prevent sybil attacks.     
+                    if (ipEndpoint.Address.IsLocal())
+                    {
+                        // Local peer: filtering is disabled unless explicitly set to true.
+                        if (this.ConnectionSettings.IpRangeFiltering == true)
+                        {
+                            if (this.PeerIsPartOfExistingGroupOfConnectedEndpoints(peerAddress))
+                            {
+                                continue;
+                            }
+                        }
+                        await this.ConnectAsync(peerAddress).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        // Remote peer: filtering is enabled unless explicitly set to false.
+                        if (this.ConnectionSettings.IpRangeFiltering != false)
+                        {
+                            if (this.PeerIsPartOfExistingGroupOfConnectedEndpoints(peerAddress))
+                            {
+                                continue;
+                            }
+                        }
+                        await this.ConnectAsync(peerAddress).ConfigureAwait(false);
+                    }
+                }
             }
         }
+
+        private bool PeerIsPartOfExistingGroupOfConnectedEndpoints(PeerAddress peerAddress)
+        {
+            IPEndPoint[] connectedEndpoints = this.ConnectorPeers.Select(x => x.PeerEndPoint).ToArray();
+            byte[] GetGroup(IPEndPoint a) => a.Address.GetGroup();
+            return connectedEndpoints.Any(a => GetGroup(a).SequenceEqual(GetGroup(peerAddress.Endpoint.MapToIpv6())));
```

Comment:
Use more local variables, use less Linq and don't define the local getgroup method.

Also at comments to the method.

And you probably don't want to use this.ConnectorPeers but this.connectedPeers, or is it something intentional?

In any case, we probably don't need to construct the list of endpoints every time for each peer in this method.

Reply:
RE: Use more local variables, use less Linq and don't define the local getgroup method.
[DONE]
RE: this.ConnectorPeers
[connectedPeers is inaccessible an doesn't reflect the connected peers immediately like ConnectorPeers]

RE: we probably don't need to construct the list of endpoints every time for each peer in this method
[This now happens only once outside the loop]